importers/src/bible/osis/index.ts

```diff
@@ -0,0 +1,945 @@
+                // sometimes there is a verse 1 tag, sometimes not. sometimes the same verse number
+                // has two verse tags when a section splits a verse
+                if (+numbers[2] !== 1 && +numbers[2] !== context.currentVerse)
+                    context.currentVerse++;
+                // sometimes there are skipped verses in versions
+                if (context.currentVerse + 1 === +numbers[2]) context.currentVerse++;
+                if (+numbers[1] !== context.currentChapter || +numbers[2] !== context.currentVerse)
+                    throw new Error(`numbering mismatch: ${tag.attributes.osisID}`);
+                break;
+            }
+            case OsisXmlNodeType.SECTION_MAJOR:
+            case OsisXmlNodeType.SECTION:
+            case OsisXmlNodeType.SECTION_SUB: {
+                // since we have some osis files where subSections are ended too early, we only
+                // close sections when the next one is started - this leads to this rather
+                // complicated code where we look which kind of section is currently open in order
+                // to determine if we close the section(s)
+
+                let currentContainer = this.getCurrentContainer(context);
+
+                // if a section starts, we close all still open groups (even if the file
+                // isn't buggy, this can happen, since we sometimes auto open certain
+                // groups)
+                while (currentContainer.type !== 'section' && currentContainer.type !== 'root') {
+                    context.contentContainerStack.pop();
+                    currentContainer = this.getCurrentContainer(context);
+                }
+                // => currentContainer is now either a section or the root node
+
+                if (context.sectionStack.length) {
+                    const sectionOrder = [
+                        OsisXmlNodeType.SECTION_MAJOR,
+                        OsisXmlNodeType.SECTION,
+                        OsisXmlNodeType.SECTION_SUB
+                    ];
+                    const elementSectionOsisLevel = sectionOrder.indexOf(elementType);
+                    let currentSectionOsisLevel = sectionOrder.indexOf(
+                        context.sectionStack[context.sectionStack.length - 1]
+                    );
+
+                    while (elementSectionOsisLevel <= currentSectionOsisLevel) {
+                        context.sectionStack.pop();
+                        // also works when sectionStack is empty
+                        currentSectionOsisLevel = sectionOrder.indexOf(
+                            context.sectionStack[context.sectionStack.length - 1]
+                        );
+
+                        context.contentContainerStack.pop();
+                        currentContainer = this.getCurrentContainer(context);
+                    }
+                }
+
+                if (currentContainer.type !== 'root' && currentContainer.type !== 'section')
```

Comment:
prefer enum or constant for string literals

Reply:
those are the biblengine container types. refactoring them to literals would be a big change